docs: add auth security doc, reorder auth navbar (#9139)
supabase/supabase: apps/reference/docs/guides/auth/auth-security.mdx

URL that contains a single-use and unguessable code. By default this URL is
hosted by the Supabase Auth server and immediately redirects back to your app.

You should be aware that email messages are inherently insecure, although the
default SMTP settings from Supabase Auth always attempt to deliver email
messages via TLS to the receiving server.

There are a number of email clients, enterprise and end-user security software,
which inspect email messages for malware. Often this software will visit links
in messages (sometimes multiple times) in an attempt to scan for viruses.
Unfortunately this has the effect of signing up a user without their knowledge
or express consent. Furthermore, it does give out an access and refresh token
to this software which could be used maliciously.

:::warning
We do not recommend using magic-link sign in without customization for
applications requiring a high level of authentication security. Should you wish
to still use magic-links you can prevent this type of issues by modifying the
magic link email template to point to your application where you can present
the user with a button that will lead to the magic link URL.

> :::

Author: TODO: Link to a guide on how to customize the magic link page.

Author: Maybe for another time, we don't have a guide for that right now.

> :::

Reviewer: I'm guessing customisation here refers to adding additional steps to the magic link workflow(e.g. verification of another factor or adding additional checks to ensure user is who they claim to be). 

Let's check with @hf though

## User information

Supabase Auth stores user account information in your project's database. This
information is provided by your application or collected from third-party
identity providers.

Email addresses, phone numbers and other personally identifiable information is
not stored encrypted in your project's database. If you need to store encrypted
data for users, we recommend you check out Supabase Vault as an easy and safe

> way to achieve this.

Reviewer: Hmm was always under the impression that Supabase vault was for secrets at the database level -- I would it work as bulk storage for individual user email and details? Imagine it'd be on the order of 100,000++ rows for big ticket users

### User enumeration protection

> User enumeration is a technique where third-parties gather information about
> the existence of accounts on a system. In and of itself the use of this
> technique does not indicate malicious behavior, but it may be used as part of
> such an effort.

Reviewer 2: ```suggestion
User enumeration is a technique where third-parties gather information about
the existence of accounts on a system. The use of this
technique does not indicate malicious behavior, but it may be used as part of
such an effort.
```

Although [industry best
practices](https://cheatsheetseries.owasp.org/cheatsheets/Authentication_Cheat_Sheet.html#authentication-and-error-messages)
ask for applications to mask the existence of accounts in their system to
unauthorized parties, this is easier said than done, especially when dealing
with a system using email and password login. Below are some examples where
applications typically fail to mask the existence of an account.

1. **Side-channel leak via password hashing.**  
Even if an application returns ambiguous messages such as "Email or password
not correct." the login request will have a significantly longer duration if
the account exists. This is because password-hashing functions take hundreds of
milliseconds to compare a stored password hash with the provided and possibly
incorrect password. A typical solution to this problem is making all login

> requests have an equal and long duration to obfuscate the fact. However, this
> is only a half-way solution since attackers can intentionally overload a system
> and still gather information from the side-channel.

Reviewer 2: I think "half-baked" is a more common expression but I could be wrong. 😅 
Let's just say "partial" to eliminate any confusion.
```suggestion
requests have an equally long duration to obfuscate the fact. However, this
is only a partial solution since attackers can intentionally overload a system
and still gather information from the side-channel.
```

2. **Side-channel leak via email or SMS.**  
With passwordless login or flows such as forgot password, information about
whether an account exists in the system can leak even if the application
returns an ambigouous message such as "If you have an account an email has been

> sent." This is because the final part of the request handling will need to send
> out an email message or SMS using a third-party system. This makes requests
> naturally last longer when an account exists. Sometimes applications add
> per-account rate-limiting to stop abuse, but this can exacerbate the problem as
> rate-limiting counts toward accounts that exist, not those that do not. Thus
> attackers can intentionally trigger rate-limits and gather information that
> way.

Reviewer 2: ```suggestion
sent." This is because the final part of the request handling needs to send an email message or SMS using a third-party system and requests last longer when an account exists. Sometimes applications add
per-account rate-limiting to stop abuse, but this can exacerbate the problem as
rate-limiting counts toward accounts that exist, not those that do not. Thus
attackers can intentionally trigger rate-limits and gather information that
way.
```

3. **Account creation vs. login.**  

> Often applications will obfuscate login requests, but not account creation
> requests, thus leaking information and confusing legitimate users. Those on the
> login flow will be told "Email or password is not correct." while on the
> sign-up flow they will be outright told "An account with this email address
> already exists." The problem compounds if the user has both social login and
> password based login. In this situation there is no solution to the problem,
> you can either: tell the user outright that there is already an account,
> redirect to the social login provider (which again leaks the information), or
> return a server error (that still leaks the information).

Reviewer 2: ```suggestion
Applications often obfuscate login requests, but not account creation
requests, which can leak information and confuse legitimate users. 
Login flows typically show a message like "Email or password is not correct," whereas sign-up flows usually show "an account with this email address already exists." The problem compounds if the user has both a social login and
a password-based login. In this situation,
you can tell the user that there is already an account and
redirect to the social login provider (which again leaks the information) or
return a server error (still leaks the information).
```

> :::tip
> Large scale user enumeration attacks can be partially mitigated or made
> inefficient by enabling some form of abuse protection on your project, such as
> setting adequate per-IP rate limiting and the introduction of a Turing test
> (CAPTCHA).
> :::

Reviewer 2: ```suggestion
Large scale user enumeration attacks can be partially mitigated or made
inefficient by enabling some form of abuse protection for your project, such as
setting adequate per-IP rate limiting or a Turing test system like
CAPTCHA.
```

> Because having a solution that covers all possible cases is virtually
> impossible, Supabase Auth gives out these conditional guarantees on the user
> enumeration problem.
> 
> We believe these guarantees strike the right balance between user experience,
> system performance, privacy and security both for hobby and enterprise
> projects.

Reviewer 2: Should we add a sub-heading for this section(e.g., "Guarantees")?
```suggestion
As having a solution that covers all possible cases is virtually
impossible, Supabase Auth gives out these conditional guarantees on the user
enumeration problem.

We believe these guarantees strike the right balance between user experience,
system performance, privacy, and security, both for hobby and enterprise
projects.
```

**No guarantees.**

> Supabase Auth is likely to leak information about the existence of any account
> that has an associated password and the password is a viable login mechanism
> for that account. Furthermore, to make it easier for users Supabase Auth may
> directly communicate with users about the existence of their account and/or
> correctness of their password.
> 
> Supabase Auth is also likely to leak information about the existence of any
> account through a side-channel if any email or phone passwordless
> authentication is used. Furthermore, to make it easier for users Supabase Auth
> may directly communicate with users about the existence of their account, any
> blocking measures and/or rate-limiting applied to their account.

Reviewer 2: ```suggestion
Supabase Auth is likely to leak information about the existence of any account
that has an associated password and the password is a viable login mechanism
for that account. To make it easier for users, Supabase Auth may directly communicate with users about the existence of their account or
correctness of their password.

Supabase Auth is also likely to leak information about the existence of any
account through a side-channel if email or phone passwordless
authentication is used. To make it easier for users, Supabase Auth
may directly communicate with users about the existence of their account, any
blocking measures, or rate-limiting applied to their account.
```

**Weak guarantees.**

> Supabase Auth is not likely to divulge information about the existence of
> any account in the system provided the account is accessible only via one and
> only one social login provider, and no other social login providers are
> enabled for the project.
> 
> Any attempts to use multiple social login providers for an account, especially
> ones that do not offer guarantees about the verification status of the email
> address they advertise, may lead to leaking of information about the existence
> or blocking measures applied to a user account.

Reviewer 2: ```suggestion
Supabase Auth is not likely to divulge information about the existence of
any account in the system provided the account is accessible through one and
only one social login provider, and no other social login providers are
enabled for the project.

Any attempts to use multiple social login providers for an account, especially
ones that do not offer guarantees about the verification status of the email
address, may lead to leaking of information about the existence
or blocking measures applied to a user account.
```

**Stronger guarantees.**

> Supabase Auth will not leak information about users with account accessible
> through enterprise-level identity providers via the SAML or OIDC protocols,
> provided that no account linking has taken place on those accounts with social
> login providers or with password and passwordless authentication.
> 
> :::tip
> If your application has high privacy requirements, we recommend asking your
> users to obfuscate their email address by using virtual email addresses (by
> using the form `user+obfuscation@example.com` if allowed by their email address
> provider),  or use a trusted social login provider.
> :::

Reviewer 2: ```suggestion
Supabase Auth will not leak information about users with accounts accessible
through enterprise-level identity providers via the SAML or OIDC protocols,
provided that no account linking has taken place on those accounts with social
login providers or with password and passwordless authentication.

If your application has high privacy requirements, your
users should obfuscate their email address by using virtual email addresses (e.g., `user+obfuscation@example.com` if allowed by their email address
provider) or use a trusted social login provider.
```

---
id: auth-security
title: 'Security'
description: Important information about security and privacy when using Supabase Auth.
---

Allowing you to build secure applications with Supabase is our priority. This guide sheds some light on the measures we take to secure
Supabase Auth and provides some guidance on building secure applications.

:::note
To report a security vulnerability with Supabase Auth or any
other Supabase service, email [security@supabase.com](mailto:security@supabase.com). Review the

> [security.txt](https://supabase.com/.well-known/security.txt) policy for a list of vulnerabilities that are not in scope.

Reviewer 3: ```suggestion
[security.txt](https://supabase.com/.well-known/security.txt) policy for a list of vulnerabilities that are in scope.
```

:::

## Credentials

> Credentials such as email addresses, phone numbers, passwords, and one-time
> passwords are always sent to Supabase Auth as `POST` requests over TLS. This
> prevents them from being accidentally logged in HTTP request logs as part of
> the URL.

Reviewer 3: ```suggestion
Credentials such as email addresses, phone numbers, passwords, and one-time
passwords are always sent to Supabase Auth as `POST` requests over TLS. This
prevents them from being logged as part of the URL in HTTP request logs.
```

An exception to this rule is magic links, which encode a short-lived,
difficult-to-guess code in a URL. This URL sends a `GET` request to
Supabase Auth and the code can only be used once.

Passwords are stored in your project's database hashed with the
[bcrypt](https://en.wikipedia.org/wiki/Bcrypt) password hashing function at a
difficulty level of 10.

Unlike other authentication services, Supabase Auth stores each
project's user information in isolated Postgres databases. This means that
security incidents involving other projects will not result in an incident in
your project.

> Social login (e.g., Login with GitHub) often uses the OAuth authorization

Reviewer 3: ```suggestion
Social logins (e.g., Login with GitHub) often uses the OAuth authorization
```

protocol. Supabase Auth does not store the token issued by the provider, but

> does make it available to your browser's app should you need further access to

Reviewer 3: ```suggestion
makes it available to your browser's app if you need further access to
```

per-account rate-limiting to stop abuse, but this can exacerbate the problem as
rate-limiting counts toward accounts that exist, not those that do not. Thus
attackers can intentionally trigger rate-limits and gather information that
way.
3. **Account creation vs. login.**  
Often applications will obfuscate login requests, but not account creation
requests, thus leaking information and confusing legitimate users. Those on the
login flow will be told "Email or password is not correct." while on the
sign-up flow they will be outright told "An account with this email address
already exists." The problem compounds if the user has both social login and
password based login. In this situation there is no solution to the problem,
you can either: tell the user outright that there is already an account,
redirect to the social login provider (which again leaks the information), or
return a server error (that still leaks the information).

Large scale user enumeration attacks can be partially mitigated or made
inefficient by enabling some form of abuse protection for your project, such as
setting adequate per-IP rate limiting or a Turing test system like
CAPTCHA.

### Guarantees

> As having a solution that covers all possible cases is virtually impossible,
> Supabase Auth gives out these conditional guarantees on the user enumeration
> problem.
> 
> We believe these guarantees strike the right balance between user experience,
> system performance, privacy and security both for hobby and enterprise
> projects.

Reviewer 3: ```suggestion
It's difficult to give consistent guarantees across all users and authentication methods. Supabase Auth provides conditional guarantees which strike the right balance between user experience,
system performance, privacy, and security both for hobby and enterprise projects.
```

**No guarantees.**

Supabase Auth is likely to leak information about the existence of any account
that has an associated password and the password is a viable login mechanism

> for that account. To make it easier for users, Supabase Auth may directly communicate with users about the existence of their account or

Reviewer: This was a little hard to digest on first pass but I can't think of a way to paraphrase offhand either because an accounts can have password as a viable login mechanism but not have an associated password

Reviewer 3: I tried to re-write this, but I actually don't know what it means. 

"Supabase Auth may directly communicate with users" - is this Supabase the company? 

"Supabase Auth is likely to leak information about the existence of any account
that has an associated password and the password is a viable login mechanism
for that account." - what is getting leaked? This sounds scary

Author: Will rephrase. "Leak information about the existence of any account" -- give an indication to unauthenticated users that an account exists.

account in the system provided the account is accessible through one and only
one social login provider, and no other social login providers are enabled for
the project.

Any attempts to use multiple social login providers for an account, especially
ones that do not offer guarantees about the verification status of the email
address, may lead to leaking of information about the existence or blocking
measures applied to a user account.

**Stronger guarantees.**

Supabase Auth will not leak information about users with accounts accessible
through enterprise-level identity providers via the SAML or OIDC protocols,
provided that no account linking has taken place on those accounts with social
login providers or with password and passwordless authentication.

If your application has high privacy requirements, your
users should obfuscate their email address by using virtual email addresses (e.g., `user+obfuscation@example.com` if allowed by their email address
provider) or use a trusted social login provider.

## Access and refresh tokens

A successful user login triggers Supabase Auth to issue an access and refresh
token.

> Access tokens are in the form of a JSON Web Token (JWT) and include most
> standard JWT claims for this purpose. At this time each JWT is signed using the
> `HS256` algorithm (HMAC with SHA256). The secret key used to derive the
> signature is generated from a cryptographically-secure pseudo random number
> generator when a project is initially created and is 512 bits long. This key is
> stored encrypted in our database, so Supabase staff are unable to readily
> access it.

Reviewer 3: ```suggestion
Access tokens are in the form of a JSON Web Token (JWT) and include most
standard JWT claims for this purpose. Each JWT is signed using the
`HS256` algorithm (HMAC with SHA256). The secret key used to derive the
signature is generated from a 512 bit, cryptographically-secure pseudo random number
generator when a project is initially created. This key is
stored encrypted in our database, so Supabase staff are unable to readily
access it.
```

We believe in giving project owners maximum flexibility, so this key is visible
to you only in your project's dashboard. Make sure you never share this value
with parties you don't completely trust.

:::tip

> You can always rotate the JWT signing key if you suspect it has leaked.
> Rotating keys on a regular basis is a good practice. Note that rotating a key
> means that all access tokens will immediately become invalid. Make sure you do
> not forget to update all other services that you use Supabase Auth access
> tokens in with the rotated secret to avoid any disruptions.

Reviewer 3: ```suggestion
You can rotate the JWT signing key if you suspect it has leaked.
Rotating keys on a regular basis is a good practice. Rotating a key
means that all access tokens will immediately become invalid. 
Remember to update all services that use Supabase Auth access
tokens with the rotated secret.
```

:::

> Access tokens are also short lived, and you can configure their validity
> duration in the project's settings page. The default value at this time is 1
> hour, which is common across the industry. We do not recommend exceeding this
> duration period, and recommend using a shorter value if it makes sense for your
> application. Durations less than 5 minutes are discouraged as computers often
> may have significant clock skew that could introduce confusing and difficult to
> debug application errors and behaviors. Shortening the duration also increases
> the application's dependence of the uptime of the authentication server.

Reviewer 3: ```suggestion
Access tokens are short lived, and you can configure their validity
duration in the project's settings page. The default value is 1
hour, which is common across the industry. We do not recommend exceeding this, 
and recommend using a shorter value if it is sensible for your
application. Durations of less than 5 minutes are discouraged as computers
may have significant clock skew that introduce confusing application errors and behaviors 
which are difficult to debug. Shortening the duration also increases
the application's dependence of the uptime of the authentication server.
```

> The purpose of access tokens is to be readily shared between components of your
> application, be it the browser, Supabase services such as PostgREST and
> functions, your hosting or other API servers under your control. Due to
> their short lifetime, any security incident involving them will have a
> time-limited blast radius.

Reviewer 3: ```suggestion
Access tokens can be readily shared between components of your
application, including the browser, Supabase services (such as PostgREST and
Edge Functions), or other servers that you control. Due to
their short lifetime, any security incident involving access tokens will have a
time-limited blast radius.
```

> Always make sure that access tokens are passed as headers or cookies, and avoid
> passing them as URL query parameters (regardless of request methods) to prevent
> them from appearing in access logs which may not be under your total control.
> When using them in Supabase functions or other API servers, take care to not
> accidentally write them to logs.

Reviewer 3: ```suggestion
Always pass access tokens as headers or cookies rather than
URL query parameters (regardless of request methods) to prevent
them from appearing in access logs.
When using access tokens in Edge Functions or other API servers, take care to not
accidentally write them to logs.
```

> It is for this reason that Supabase Auth returns both the access token and
> refresh token to your browser application via the URL fragment (anything after
> the `#` sign), as browsers will actually avoid sending this data to the
> underlying server. Thus this makes sure that the access and refresh token are
> received only by the JavaScript running in your browser and does not end up in
> access logs or is visible by your application's hosting service. Even if the
> servers hosting your browser application are under your total control, we've
> found that often there are various other appliances that intercept HTTP
> requests where disabling logging is difficult or sometimes impossible.

Reviewer 3: ```suggestion
Supabase Auth returns the access token and
refresh token to your browser application via the URL fragment (anything after
the `#` sign). This prevents browsers sending the data to the
server. URL fragments are only
accessible to the JavaScript running in your browser.
```

I removed a chunk from this one as it's been repeated a few times and it feels like this is a big vague: "e've
found that often there are various other appliances that intercept HTTP
requests where disabling logging is difficult or sometimes impossibl"

> Access tokens are not stored by Supabase Auth or other services, regardless if
> they are valid or not. We also recommend that any Supabase functions or API
> servers you use avoid storing access tokens they've seen.

Reviewer 3: ```suggestion
Access tokens are never stored by Supabase Auth or other services, regardless if
they are valid or not. We recommend that you avoid storing  them in Edge Functions or  otherAPI
servers.
```

Because of these common cases, rate-limit protection per IP address is quite
generous but is not effective at distributed attacks.

Supabase Auth does support using a [Turing
test](https://en.wikipedia.org/wiki/CAPTCHA) in the form of a CAPTCHA via the
[hCaptcha](https://www.hcaptcha.com) service to identify humans vs. robots. You
can enable this protection in your project's dashboard.

:::tip
The [W3C does not recommend the use of
CAPTCHA](https://www.w3.org/TR/turingtest/) or other Turing tests for most
web applications as they pose challenges with accessibility.
:::

## Redirect URLs

Supabase Auth lets you redirect back to different URLs on susccessful
authentication. Your client app can specify the URL to redirect back to using
the `redirectTo` parameter.

Given that in sophisticated attacks this parameter may come from a malicious
source, Supabase Auth implements an allow list which will be consulted prior to
sending a redirect response. Should the redirect parameter not fall in the URL
allow list, the Site URL is used.

The allow list can be a simple list of exactly matching URLs, but it also
supports wildcard characters to match them as a pattern.

:::tip

> You can configure the URL allow list in the [Authentcation General Settings

Reviewer: /nit typo
```suggestion
You can configure the URL allow list in the [Authentication General Settings
```
